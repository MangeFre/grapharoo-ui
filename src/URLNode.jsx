import React from 'react';
import { toast } from 'react-toastify';
import HTML from 'html-parse-stringify';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faArrowUp } from '@fortawesome/free-solid-svg-icons';
import { faArrowDown } from '@fortawesome/free-solid-svg-icons';

toast.configure();

function unescapeHTML(escapedHTML) {
	return escapedHTML
		.replace(/&lt;/g, '<')
		.replace(/&gt;/g, '>')
		.replace(/&amp;/g, '&');
}

function traverseJsonRec(node, options) {
	// If this node is a tag (a, p, etc) and the options want to modify that part.
	if (node.type === 'text') {
		return;
	}

	if (node.type === 'tag' && options[node.name]) {
		const allKeys = Object.keys(options[node.name]);
		for (let key of allKeys) {
			node.attrs[key] = options[node.name][key];
		}
	}

	for (let child of node.children) {
		traverseJsonRec(child, options);
	}

	return;
}

function addOptionsToHTML(unescapedHTML, options = {}) {
	let newHTML = unescapedHTML;
	let newJson = HTML.parse(newHTML);
	traverseJsonRec(newJson[0], options);
	newHTML = HTML.stringify(newJson);
	return newHTML;
}

function getFormattedTimePassed(created_utc) {
	let output = '';
	let created = Date.parse(created_utc) / 1000; // convert created_utc to Date and get it in seconds
	let now = Date.now() / 1000; // get current time in seconds
	let timeDiff = now - created;
	let minutes = timeDiff / 60;
	let hours = minutes / 60;
	let days = hours / 24;
	let weeks = days / 7;
	let months = weeks / 4;
	let years = days / 365;

	if (minutes < 0) {
		output = 'just now';
	} else if (minutes < 60) {
		output = `${Math.floor(minutes)} minute${minutes >= 2 ? 's' : ''} ago`;
	} else if (hours < 24) {
		output = `${Math.floor(hours)} hour${hours >= 2 ? 's' : ''} ago`;
	} else if (days < 7) {
		output = `${Math.floor(days)} day${days >= 2 ? 's' : ''} ago`;
	} else if (weeks < 4) {
		output = `${Math.floor(weeks)} week${weeks >= 2 ? 's' : ''} ago`;
	} else if (months < 12) {
		output = `${Math.floor(months)} month${months >= 2 ? 's' : ''} ago`;
	} else {
		output = `${Math.floor(years)} year${years >= 2 ? 's' : ''} ago`;
	}
	return output;
}

function getFormattedScore(score, score_hidden) {
	return score_hidden
		? '[score hidden]'
		: `${score} point${score !== 1 ? 's' : ''}`;
}

export default function URLNode(props) {

	const { url, comment, hasError } = props;

	// Conditionally changing CSS if error.
	return (
		<>
			<div className="container">
				<div className="votesContainer">
					<div className="vote up">
						<FontAwesomeIcon icon={faArrowUp}></FontAwesomeIcon>
					</div>
					<div className="vote down">
						<FontAwesomeIcon icon={faArrowDown}></FontAwesomeIcon>
					</div>
				</div>
				{comment ? 
					// Has Comment
					<div>
						<div className="headerContainer">
							{comment.author !== '[deleted]' ? (
								<a
									className="author"
									style={{ textDecoration: 'none' }}
									href={`https://www.reddit.com/user/${comment.author}`}
									target="blank">
									{comment.author}
								</a>
							) : (
								<div style={{ color: '#888' }}>{comment.author}</div>
							)}
							<div className="details">
								<div>{getFormattedScore(comment.score, comment.score_hidden)}</div>
								<div>{getFormattedTimePassed(comment.created_utc)}</div>
								<div>
									in{' '}
									<a href={url} target="blank">
										{comment.subreddit_name_prefixed}
									</a>
								</div>
							</div>
						</div>
						<p
							dangerouslySetInnerHTML={{
								__html: addOptionsToHTML(unescapeHTML(comment.body_html), {
									a: { target: 'blank' },
								}),
							}}></p>
					</div>
					:
					// Has no comments
					<div>
						<div className="headerContainer">
							<div className="loading">
								{hasError ? "Error Loading" : "Loading"}
							</div>
						</div>
						<p className="loading" >
							<a href={url} target="blank">
								{url}
							</a>
						</p>
					</div>
				}
			</div>
			<style jsx>{`
				.container {
					display: flex;
					flex-direction: column;
					margin: 1rem 1rem 0rem 1rem;

					display: flex;
					flex-direction: row;

					background-color: white;

					border-radius: 0.5rem;

					-webkit-box-shadow: 5px 5px 9px 3px rgba(0, 0, 0, 0.38);
					-moz-box-shadow: 5px 5px 9px 3px rgba(0, 0, 0, 0.38);
					box-shadow: 5px 5px 9px 3px rgba(0, 0, 0, 0.38);
				}

				.container:last-of-type {
					margin-bottom: 1rem;
				}

				.container > div {
					padding: 0.5rem;
				}

				.headerContainer {
					display: flex;
					font-size: 1.25rem;
					font-weight: 900;
					color: #336699;
				}

				.headerContainer .details {
					padding: 0 0.75em;
					color: #888;
					font-size: 0.75;

					display: flex;
				}

				.headerContainer .details div {
					padding-right: 0.25rem;
				}

				.headerContainer .loading {
					color: black;
					padding-bottom: 0.75em;
				}

				.votesContainer {
					display: flex;
					flex-direction: column;
				}

				.vote {
					color: #c6c6c6;
					font-size: 1.5rem;
				}

				.up {
					color: #ff8b60;
				}

				.container:last-of-type .up {
					color: #c6c6c6;
				}

				.container:last-of-type .down {
					color: #9494ff;
				}

				.info {
					display: flex;
				}

				p {
					padding: 0;
					font-size: 1.5rem;
				}

				h2,
				p {
					margin: 0;
				}

				p.loading {
					padding-bottom: 0.75em;
				}
			`}</style>
		</>
	);
}
