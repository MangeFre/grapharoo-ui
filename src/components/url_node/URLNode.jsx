import React, { useState, useEffect } from 'react';

import { toast } from 'react-toastify';

import BrokenComment from '../reddit_comment/BrokenComment';
import Comment from '../reddit_comment/Comment';
import LoadingComment from '../reddit_comment/LoadingComment';
import VoteArrows from '../reddit_comment/VoteArrows';

toast.configure();

export default function URLNode(props) {

	const { 
		url,
		errorUrl, 
		loadingUrl,
		comment,  
		setFixedLink, 
		seen,
		fixed
		
	} = props;

	const [isUpvoted, setIsUpvoted] = useState(false);
	const [isDownvoted, setIsDownvoted] = useState(false);

	useEffect(()=>{
		setIsUpvoted(comment !== null && comment !== undefined);
	}, [comment]);

	useEffect(()=>{
		setIsDownvoted(errorUrl !== null && errorUrl !== undefined);
	}, [errorUrl]);

	return (
		<>
			<div className='container'>
				<div className='reddit-comment'>
					<div className='vote-arrows'>
						<VoteArrows isUpvoted={isUpvoted} isDownvoted={isDownvoted}></VoteArrows>
					</div>
					<div className='comment'>
						{/*
						 Only one of the Comment elements below should be shown at a time
						*/}
						{ comment ? <Comment comment={comment} url={url} seen={seen} fixed={fixed}></Comment> : <></> }
						{ errorUrl ? <BrokenComment url={errorUrl} setFixedLink={setFixedLink}></BrokenComment> : <></> }
						{ loadingUrl ? <LoadingComment url={loadingUrl}></LoadingComment> : <></> }
					</div>
				</div>
			</div>
			<style jsx>{`
				.container {
					display: flex;
					flex-direction: row;

					margin: 1rem 1rem 0rem 1rem;
					padding: 1rem 1rem 0rem 1rem;

					border-radius: 0.5rem;

					background-color: white;

					-webkit-box-shadow: 5px 5px 9px 3px rgba(0, 0, 0, 0.38);
					-moz-box-shadow: 5px 5px 9px 3px rgba(0, 0, 0, 0.38);
					box-shadow: 5px 5px 9px 3px rgba(0, 0, 0, 0.38);
				}

				.reddit-comment {
                    display: flex;
					flex-direction: row;
                }
				
                .vote-arrows {
                    padding-right: 1em;
                }
			`}</style>
		</>
	);
}
