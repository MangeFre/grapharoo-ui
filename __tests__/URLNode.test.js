import React from 'react';
import renderer from 'react-test-renderer';
import URLNode from '../src/components/url_node/URLNode';

const url = 'https://www.reddit.com/r/aww/comments/hd6xtp/comment/fvk5vao';
const onFinishLoad = () => {};

test('Create snapshot for future comparison', () => {
	const component = renderer.create(
		<URLNode url={url} key={url} onFinishLoad={onFinishLoad} />,
	);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});
