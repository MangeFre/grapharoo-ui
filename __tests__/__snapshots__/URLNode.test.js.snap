// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create snapshot for future comparison 1`] = `
<div
  className="jsx-2394042114 container"
>
  <div
    className="jsx-2394042114 votesContainer"
  >
    <div
      className="jsx-2394042114 vote up"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-arrow-up fa-w-14 "
        data-icon="arrow-up"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M34.9 289.5l-22.2-22.2c-9.4-9.4-9.4-24.6 0-33.9L207 39c9.4-9.4 24.6-9.4 33.9 0l194.3 194.3c9.4 9.4 9.4 24.6 0 33.9L413 289.4c-9.5 9.5-25 9.3-34.3-.4L264 168.6V456c0 13.3-10.7 24-24 24h-32c-13.3 0-24-10.7-24-24V168.6L69.2 289.1c-9.3 9.8-24.8 10-34.3.4z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
    <div
      className="jsx-2394042114 vote down"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-arrow-down fa-w-14 "
        data-icon="arrow-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </div>
  <div
    className="jsx-2394042114"
  >
    <div
      className="jsx-2394042114 headerContainer"
    >
      <div
        className="jsx-2394042114 loading"
      >
        Loading
      </div>
    </div>
    <p
      className="jsx-2394042114 loading"
    >
      <a
        className="jsx-2394042114"
        href="https://www.reddit.com/r/aww/comments/hd6xtp/comment/fvk5vao"
        target="blank"
      >
        https://www.reddit.com/r/aww/comments/hd6xtp/comment/fvk5vao
      </a>
    </p>
  </div>
</div>
`;
